cache/sharded_cache.h

```diff
@@ -69,55 +76,230 @@ class CacheShard {
 // Generic cache interface which shards cache by hash of keys. 2^num_shard_bits
 // shards will be created, with capacity split evenly to each of the shards.
 // Keys are sharded by the highest num_shard_bits bits of hash value.
+// There are three template parameters:
+// - T: The type of the hashes.
+// - Hasher: Implements inline static T hash(const Slice& s). This
+//    function computes a hash, provided a key.
+// - ShardExtractor: Implements
+//    inline static uint32_t extract(T hash, uint32_t shard_mask).
+//    This function computes the index of a shard, given a hash and the shard
+//    mask.
+template <typename T, typename Hasher, typename ShardExtractor>
```

Comment:
I would prefer a name like `HashType` to `T` because it's not obvious what role the type parameter is playing.

Also, I would generally prefer either (a) group related templated functionality into members of one type parameter, e.g. Hasher::Type + Hasher::ComputeHash + Hasher::ExtractShard, or (b) use an idiom for function-like type parameters, probably with `operator()` as in https://github.com/facebook/rocksdb/blob/main/table/block_based/block.cc#L585

Google C++ style calls for function names to use initial caps, which is the same convention for user-defined types. So in this context where we have functions and types, a name like `Hash` is neither clearly a function (verb) nor clearly a type (noun).

Reply:
Ah! After doing some research, I learned that it's possible to extract the hash type from a class in a clean way. That was stopping me from bundling everything.